File: app/src/main/java/io/karpilabs/simplemp3/data/jellyfin/JellyfinModels.kt
package io.karpilabs.simplemp3.data.jellyfin

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
data class AuthenticateByNameRequest(
    @Json(name = "Username") val username: String,
    @Json(name = "Pw") val password: String
)

@JsonClass(generateAdapter = true)
data class AuthenticationResult(
    @Json(name = "AccessToken") val accessToken: String?,
    @Json(name = "User") val user: JellyfinUser?,
    @Json(name = "ServerId") val serverId: String? = null
)

@JsonClass(generateAdapter = true)
data class JellyfinUser(
    @Json(name = "Id") val id: String,
    @Json(name = "Name") val name: String? = null
)

@JsonClass(generateAdapter = true)
data class QueryResult(
    @Json(name = "Items") val items: List<JellyfinItem> = emptyList(),
    @Json(name = "TotalRecordCount") val totalRecordCount: Int = 0,
    @Json(name = "StartIndex") val startIndex: Int = 0
)

@JsonClass(generateAdapter = true)
data class JellyfinItem(
    @Json(name = "Id") val id: String,
    @Json(name = "Name") val name: String? = null,
    @Json(name = "Type") val type: String? = null,
    @Json(name = "Album") val album: String? = null,
    @Json(name = "AlbumId") val albumId: String? = null,
    @Json(name = "AlbumArtist") val albumArtist: String? = null,
    @Json(name = "Artists") val artists: List<String>? = null,
    @Json(name = "RunTimeTicks") val runTimeTicks: Long? = null,
    @Json(name = "IndexNumber") val indexNumber: Int? = null,
    @Json(name = "ProductionYear") val productionYear: Int? = null,
    @Json(name = "ImageTags") val imageTags: Map<String, String>? = null,
    @Json(name = "AlbumPrimaryImageTag") val albumPrimaryImageTag: String? = null,
    @Json(name = "ParentId") val parentId: String? = null,
    @Json(name = "ChildCount") val childCount: Int? = null,
    @Json(name = "Size") val size: Long? = null,
    @Json(name = "Container") val container: String? = null
) {
    val durationMs: Long
        get() = (runTimeTicks ?: 0L) / 10_000L

    val artistName: String
        get() = artists?.firstOrNull()?.takeIf { it.isNotBlank() }
            ?: albumArtist?.takeIf { it.isNotBlank() }
            ?: "Unknown Artist"

    val albumName: String
        get() = album?.takeIf { it.isNotBlank() } ?: "Unknown Album"

    val title: String
        get() = name?.takeIf { it.isNotBlank() } ?: "Unknown Title"

    val hasPrimaryImage: Boolean
        get() = imageTags?.containsKey("Primary") == true || !albumPrimaryImageTag.isNullOrBlank()
}

data class JellyfinSession(
    val serverUrl: String,
    val accessToken: String,
    val userId: String,
    val userName: String,
    val serverId: String? = null,
    val deviceId: String
)

data class SyncProgress(
    val phase: String = "Idle",
    val current: Int = 0,
    val total: Int = 0,
    val currentTitle: String = "",
    val isActive: Boolean = false,
    val error: String? = null,
    val lastResult: String? = null
) {
    val fraction: Float
        get() = if (total <= 0) 0f else (current.toFloat() / total).coerceIn(0f, 1f)
}

File: ios/Simple MP3/Jellyfin/JellyfinModels.swift
//
//  JellyfinModels.swift
//  Simple MP3
//

import Foundation

nonisolated struct AuthenticateByNameRequest: Encodable {
    let Username: String
    let Pw: String
}

nonisolated struct AuthenticationResult: Decodable {
    let AccessToken: String?
    let User: JellyfinUser?
    let ServerId: String?
}

nonisolated struct JellyfinUser: Decodable {
    let Id: String
    let Name: String?
}

nonisolated struct QueryResult: Decodable {
    let Items: [JellyfinItem]?
    let TotalRecordCount: Int?
    let StartIndex: Int?
}

nonisolated struct JellyfinItem: Decodable, Identifiable, Hashable {
    let Id: String
    let Name: String?
    let itemType: String?
    let Album: String?
    let AlbumId: String?
    let AlbumArtist: String?
    let Artists: [String]?
    let RunTimeTicks: Int64?
    let IndexNumber: Int?
    let ProductionYear: Int?
    let ImageTags: [String: String]?
    let AlbumPrimaryImageTag: String?
    let ParentId: String?
    let ChildCount: Int?
    let Size: Int64?
    let Container: String?

    enum CodingKeys: String, CodingKey {
        case Id, Name, Album, AlbumId, AlbumArtist, Artists
        case RunTimeTicks, IndexNumber, ProductionYear, ImageTags
        case AlbumPrimaryImageTag, ParentId, ChildCount, Size, Container
        case itemType = "Type"
    }

    var id: String { Id }

    var durationMs: Int64 { (RunTimeTicks ?? 0) / 10_000 }

    var artistName: String {
        if let a = Artists?.first, !a.isEmpty { return a }
        if let a = AlbumArtist, !a.isEmpty { return a }
        return "Unknown Artist"
    }

    var albumName: String {
        if let a = Album, !a.isEmpty { return a }
        return "Unknown Album"
    }

    var title: String {
        if let n = Name, !n.isEmpty { return n }
        return "Unknown Title"
    }

    var hasPrimaryImage: Bool {
        ImageTags?["Primary"] != nil || !(AlbumPrimaryImageTag ?? "").isEmpty
    }

    /// Convenience builder for screenshot / preview fixtures.
    static func demo(
        id: String,
        title: String,
        artist: String,
        album: String,
        durationMs: Int64
    ) -> JellyfinItem {
        // Decode from a minimal JSON payload so we don't fight Decodable-only fields.
        let json: [String: Any] = [
            "Id": id,
            "Name": title,
            "Type": "Audio",
            "Album": album,
            "Artists": [artist],
            "AlbumArtist": artist,
            "RunTimeTicks": durationMs * 10_000
        ]
        let data = try! JSONSerialization.data(withJSONObject: json)
        return try! JSONDecoder().decode(JellyfinItem.self, from: data)
    }
}

struct JellyfinSession: Equatable, Sendable {
    var serverUrl: String
    var accessToken: String
    var userId: String
    var userName: String
    var serverId: String?
    var deviceId: String
}

struct SyncProgress: Equatable, Sendable {
    var phase: String = "Idle"
    var current: Int = 0
    var total: Int = 0
    var currentTitle: String = ""
    var isActive: Bool = false
    var error: String?
    var lastResult: String?

    var fraction: Double {
        guard total > 0 else { return 0 }
        return min(1, Double(current) / Double(total))
    }
}
